0024-pre-shared-key-v2/0024-pre-shared-key-v2.md

```diff
@@ -0,0 +1,287 @@
+4. Sender sends a payment using the source amount determined in step 2, the condition from step 3, and the encrypted request as the data.
+5. Receiver is notified of the incoming transfer and decrypts the data.
+6. Receiver checks to ensure that the incoming transfer amount is greater than or equal to the Chunk Amount specified in the request. If the Payment Amount is not equal to 0, the receiver checks that the transfer amount is less than or equal to the Payment Amount specified in the request.
+7. Receiver creates a [Payment Chunk Response](#payment-chunk-response) with the same Payment ID and Sequence number. The Payment Amount and Chunk Amount are set to the amount that arrived on the incoming transfer. The receiver then [encrypts](#encryption-envelope) the data.
+8. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer with the fulfillment and the encrypted response data.
+9. Sender is notified of the outgoing transfer being fulfilled and can use the decrypted response to determine the amount that arrived at the destination.
+
+### Chunked Payment
+
+1. Sender creates a packet with Type 0, a random Payment ID, a Sequence number of 0 and a Chunk Amount of 0. In the case of a fixed destination amount, sender sets the Payment Amount to the destination amount. In the case of a fixed source amount, sender sets the Payment Amount to 2^64 - 1 (the maximum unsigned 64-bit integer, or 18446744073709551615). Sender [encrypts the request](#encryption-envelope).
+2. Sender generates the payment condition as described [below](#fulfillable-condition).
+3. Sender sends a payment with the condition from step 2., the encrypted request from step 1. as the data, and an arbitrary source amount.
+4. Receiver is notified of the incoming transfer and decrypts the data.
+5. Receiver checks that the incoming transfer amount is greater than or equal to the Chunk Amount and that the total that has arrived thus far does not exceed the Payment Amount.
+6. Receiver creates a response packet with the Type 2, the same Payment ID, Sequence. The Payment Amount is set to the total the receiver has received and fulfilled thus far (including the current chunk). Receiver sets the Chunk Amount to the amount that arrived on the incoming transfer. Receiver [encrypts the response](#encryption-envelope).
+7. Receiver generates fulfillment as described [below](#fulfillment-generation) and fulfills the incoming transfer.
+8. Sender is notified of the outgoing transfer being fulfilled and decrypts the attached data.
+9. Sender determines the exchange rate by dividing the Chunk Amount from the receiver by the chunk's source amount.
+10. Sender repeats step 1-8 with the following modifications, until the fixed source or destination amount has been sent or delivered, respectively:
+  - Sender MUST create a new PSK data packet for each chunk with the same Payment ID and an incremented Sequence number
```

Comment:
What's the purpose of the sequence number? What should the receiver do if they miss a sequence number? If one of the payments fails, should the sender increment the sequence number, or retry it with the same sequence number? Will an informational quote always have a sequence number of 0, even if it occurs partway through a payment (for example to refresh the exchange rate)?

Reply:
The main purpose of the sequence number right now is to allow the sender to ensure that responses match the requests they sent out. Without the sequence, a connector could attach an old response (that would say a lower amount had arrived) to a later fulfillment and the sender wouldn't know the difference.

Originally I thought that receivers should keep track of the sequence numbers to avoid accepting two with the same sequence. @justmoon argued that wasn't necessary because unlike with data, money is fungible so it doesn't matter if the chunks arrive out of order as long as you get the total amount.

I don't think informational quotes need to have a sequence of 0. From the receiver's perspective, there is no such thing as an informational quote, it's just a chunk that arrived with too little money (and potentially an unfulfillable condition).